pkg/registry/client.go

```diff
@@ -0,0 +1,177 @@
+package registry
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	orasregistry "oras.land/oras-go/v2/registry"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// OCIRegistryClient storage client for an OCI registry.
+type OCIRegistryClient struct {
+	host        string
+	project     string
+	certFile    string
+	insecure    bool
+	dryRun      bool
+	initialized bool
+	OI          OrasInterface
+	registry    *remote.Registry
+}
+
+var _ StorageClient = (*OCIRegistryClient)(nil)
+
+// NewOCIRegistry create an OCI registry client.
+func NewOCIRegistry(host string, certFile string, insecure bool) *OCIRegistryClient {
+	return &OCIRegistryClient{
+		host:     host,
+		certFile: certFile,
+		insecure: insecure,
+		OI:       &OrasImplementation{},
+	}
+}
+
+// Init registry configuration.
+func (or *OCIRegistryClient) Init() error {
+	if or.initialized {
+		return nil
+	}
+
+	credentialStore := NewCredentialStore()
+	err := credentialStore.Init()
+	if err != nil {
+		return err
+	}
+
+	certificates, err := or.getCertificates()
+	if err != nil {
+		return err
+	}
+
+	or.registry, err = remote.NewRegistry(or.host)
+	if err != nil {
+		return fmt.Errorf("NewRegistry: %v", err)
+	}
+
+	tlsConfig := &tls.Config{
+		RootCAs:            certificates,
+		InsecureSkipVerify: or.insecure,
+	}
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = tlsConfig
+	authClient := &auth.Client{
+		Client: &http.Client{
+			Transport: transport,
+		},
+		Cache: auth.NewCache(),
+	}
+	authClient.SetUserAgent("eksa")
+	authClient.Credential = func(ctx context.Context, s string) (auth.Credential, error) {
+		return credentialStore.Credential(s)
+	}
+	or.registry.Client = authClient
+	or.initialized = true
+	return nil
+}
+
+// GetHost for registry host.
+func (or *OCIRegistryClient) GetHost() string {
+	return or.host
+}
+
+// GetProject for registry project.
+func (or *OCIRegistryClient) GetProject() string {
+	return or.project
+}
+
+// GetCertFile for registry certificate file.
+func (or *OCIRegistryClient) GetCertFile() string {
+	return or.certFile
+}
+
+// IsInsecure insecure TLS connection.
+func (or *OCIRegistryClient) IsInsecure() bool {
+	return or.insecure
+}
+
+// SetProject for registry destination.
+func (or *OCIRegistryClient) SetProject(project string) {
+	or.project = project
+}
+
+// SetDryRun dry run validates read, but not write.
+func (or *OCIRegistryClient) SetDryRun(value bool) {
+	or.dryRun = value
+}
+
+// Destination of this storage registry.
+func (or *OCIRegistryClient) Destination(image Artifact) string {
+	return strings.Replace(image.VersionedImage(), image.Registry+"/", or.host+"/"+or.project, 1)
```

Comment:
```suggestion
	return strings.Replace(image.VersionedImage(), url.JoinPath(url.PathEscape(image.Registry), url.PathEscape(or.host), url.PathEscape(or.project)), 1)
```
It seems a little like overkill, but IIRC, some of these values were supplied by the user, and so should be considered as possibly malicious.